### PR TITLE
Issue #19 Add style checker

### PR DESCRIPTION
**Pull request by MichiruKayo**
--added check style
--fixed check style findings.

### CODE REVIEW

File: code-analysis/check-style/check-style.xml [hunk 1]
```diff
@@ -0,0 +1,109 @@
+<?xml version="1.0"?>
```
Comment (trioletas) [inline code-analysis/check-style/check-style.xml]: that's a lot of style checks..are there any presents available that we could simply reference here? just curious

File: api/src/main/java/lv/ctco/tpl/bff/integration/icndb/jokes/JokeValueModel.java [hunk 2]
```diff
@@ -6,7 +6,7 @@
 
 @Getter
 public class JokeValueModel {
-    String id;
-    String joke;
-    List<String> categories;
+    private String id;
```
Comment (trioletas) [inline api/src/main/java/lv/ctco/tpl/bff/integration/icndb/jokes/JokeValueModel.java]: can't say I like this one: _who cares_ if it's not private? It's just a stupid java bean

File: api/build.gradle [hunk 3]
```diff
@@ -24,10 +24,30 @@ apply plugin: 'org.springframework.boot'
...
+
 bootRun() {
   environment << dotenv
 }
 
+checkstyle {
+  def dir = "${parent.projectDir}/code-analysis/check-style"
+  toolVersion = '8.2'
+  configProperties.configDir = dir
+  configFile = file("$dir/check-style.xml")
```
Comment (trioletas) [inline api/build.gradle]: Can we use the default file location instead and just drop this line? I would like to reduce boilerplate configs as much as possible, let's stick with convention over configuration as much as possible please

File: api/build.gradle [hunk 4]
```diff
@@ -24,10 +24,30 @@ apply plugin: 'org.springframework.boot'
 
 apply from: 'dotenv.gradle'
 
+apply plugin: 'checkstyle'
+apply plugin: 'pmd'
+pmdTest.enabled = false
```
Comment (trioletas) [inline api/build.gradle]: why?

File: README.md [hunk 5]
```diff
@@ -29,6 +29,18 @@ Spring Boot, GraphQL template project with batteries included.
 
 - Download and install GraphQLJs [plugin](https://plugins.jetbrains.com/plugin/8097-js-graphql)
 
+### Style check
+
+#### IntelliJ
+
+- Edit configurations > "Application" > Add task in "Before launch" section "checkstyleMain pmdMain" and place it as first
```
Comment (trioletas) [inline README.md]: is necessary ? can we just use the plugin and be done with that?

File: README.md [hunk 6]
```diff
@@ -29,6 +29,18 @@ Spring Boot, GraphQL template project with batteries included.
...
+- Additionally you can download "CheckStyle plugin". Then open Settings > Other Settings > Checkstyle. Add path to check-style.xml file 
+  and provide link to check-style folder in "${configDir}". This will display errors in editor, but it wont affect build/run.
```
Comment (trioletas) [inline README.md]: > but it won affect build/run
sounds fine by me :)

File: README.md (hunk 5 above)
Comment (MichiruKayo) [reply README.md]: If you start using intellij idea Application run/debug then plugin is not executed.. couldn't find how to force it..

File: README.md [hunk 7]
```diff
@@ -29,6 +29,18 @@ Spring Boot, GraphQL template project with batteries included.
...
+
+#### IntelliJ
+
+- Edit configurations > "Application" > Add task in "Before launch" section "checkstyleMain pmdMain" and place it as first
+- Additionally you can download "CheckStyle plugin". Then open Settings > Other Settings > Checkstyle. Add path to check-style.xml file 
+  and provide link to check-style folder in "${configDir}". This will display errors in editor, but it wont affect build/run. 
+
+#### Gradle
+
+- As default behaviour it is configured to call "checkstyleMain pmdMain" before "bootRun" task is executed
```
Comment (trioletas) [inline README.md]: Not sure I like this one, it basically stops me from launching the dev mode unless all my modifiers are private and stuff.. Can we relax this one?

File: api/build.gradle [hunk 8]
```diff
@@ -24,10 +24,30 @@ apply plugin: 'org.springframework.boot'
...
 bootRun() {
   environment << dotenv
 }
 
+checkstyle {
+  def dir = "${parent.projectDir}/code-analysis/check-style"
+  toolVersion = '8.2'
+  configProperties.configDir = dir
+  configFile = file("$dir/check-style.xml")
+  sourceSets = [] // used to skip tests
```
Comment (trioletas) [inline api/build.gradle]: why?

File: api/build.gradle (hunk 4 above)
Comment (MichiruKayo) [reply api/build.gradle]: In tests we usually have hard coded strings/ magic numbers, and style check fails in such case..

File: api/build.gradle [hunk 9]
```diff
@@ -24,10 +24,30 @@ apply plugin: 'org.springframework.boot'
...
+}
+
+pmd {
+  toolVersion = '5.8.1'
+  consoleOutput = true
+  ruleSetFiles = files("${parent.projectDir}/code-analysis/pmd/pmd.xml")
```
Comment (trioletas) [inline api/build.gradle]: I think we can do a convention over configuration with this one as well. I do recall there being a default directory where pmd looks for its configs

File: code-analysis/check-style/check-style.xml [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        "-//Puppy Crawl//DTD Check Configuration 1.2//EN"
+        "http://www.puppycrawl.com/dtds/configuration_1_2.dtd">
+
+<module name="Checker">
+    <module name="SuppressionFilter">
+        <property name="file" value="${configDir}/suppressions.xml"/>
+    </module>
+
+    <module name="Header">
+        <property name="header" value="Copyright Swiss Reinsurance Company, Mythenquai 50/60, CH 8022 Zurich. All rights reserved."/>
```
Comment (trioletas) [inline code-analysis/check-style/check-style.xml]: oops you did it again

please be careful with this stuff

File: code-analysis/check-style/suppressions.xml [hunk 11]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+
+<!DOCTYPE suppressions PUBLIC
+    "-//Puppy Crawl//DTD Suppressions 1.1//EN"
+    "http://www.puppycrawl.com/dtds/suppressions_1_1.dtd">
+
+<suppressions>
+    <!--Add exclusion here (e.g <suppress files="JokeResolver.java" checks="MethodLength"/>) -->
```
Comment (trioletas) [inline code-analysis/check-style/suppressions.xml]: Are we trying to teach how to use `check-style` here? Please drop all unnecessary config files, thanks!

File: code-analysis/check-style/check-style.xml (hunk 1 above)
Comment (MichiruKayo) [reply code-analysis/check-style/check-style.xml]: There are 2 presets available : 
http://checkstyle.sourceforge.net/style_configs.html Google and Sun, will check how to include them by default

File: api/build.gradle (hunk 4 above)
Comment (trioletas) [reply api/build.gradle]: can we use inline directives to suppress the linting for such cases? tests are also code so we should treat them as such

File: README.md (hunk 5 above)
Comment (trioletas) [reply README.md]: Okay let's think about that. But please remove these instructions from the documentation. its unnecessarily cumbersome

File: code-analysis/check-style/check-style.xml (hunk 1 above)
Comment (MichiruKayo) [reply code-analysis/check-style/check-style.xml]: unfortunately it is not possible to have 2 configs . Issue is open in style checker :https://github.com/checkstyle/checkstyle/issues/991

**Comment by MichiruKayo** [comment]
1: Removed pmd checks (not worth it)
2: Based on rij checks, reviewed them and removed ones i think are not needed
3: updated read.me
4:  config location is required. Default path is <build.gradle.folder>/config/stylecheck/stylecheck.xml, in our case it is api/config/... , dont think adding such folder is good.. maybe better to change it to root build.gradle? ( but i do not know how to configure that)
5: Changed dependendOn to "build" will probably be called during deployment?.. not sure where to attach it, since we do not have many tasks..

**Comment by trioletas** [comment]
>4: config location is required. Default path is <build.gradle.folder>/config/stylecheck/stylecheck.xml, > in our case it is api/config/... , dont think adding such folder is good.. maybe better to change it to > > root build.gradle? ( but i do not know how to configure that)
I do not see a problem with the `api/config` folder. It is a configuration for the api module is it not?

**Comment by MichiruKayo** [comment]
Changed to default location as we discussed offline

**Comment by trioletas** [comment]
Thanks, looks good to me.
@sergeytrasko any comments from your side? Otherwise we're good to go.

**Pull request merged**